File: java/leetcode/hashtable/two_sum.java
/**
 * 1. Two Sum
 * Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:

Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
 */

import java.util.HashMap;

public class two_sum {

    // sort the array and use binary search two iterators. time O(NlogN), space O(1)
    // use HashMap find nums[i] and target - nums[i], time O(N), space O(N)
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> indexForNum = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (indexForNum.containsKey(target - nums[i])) {
                return new int[]{indexForNum.get(target-nums[i]), i};
            } else {
                indexForNum.put(nums[i], i);
            }
        }
        return null;
    }
}

File: golang/leetcode/hashmaps/two_sum.go
package hashmaps

// 1. Two Sum

func twoSum(nums []int, target int) []int {
	index := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		if _, ok := index[target-nums[i]]; ok {
			return []int{index[target-nums[i]], i}
		} else {
			index[nums[i]] = i
		}
	}
	return nil
}
